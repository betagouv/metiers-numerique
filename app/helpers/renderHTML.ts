import { handleError } from '@common/helpers/handleError'
import React from 'react'
import rehypeParse from 'rehype-parse'
import rehypeReact from 'rehype-react'
import { unified } from 'unified'

import type { JSXElementConstructor, ReactElement } from 'react'

export function renderHTML(HTMLSource: string): ReactElement<unknown, string | JSXElementConstructor<any>> | null {
  try {
    return unified().use(rehypeParse).use(rehypeReact, { createElement: React.createElement }).processSync(HTMLSource)
      .result
  } catch (err) /* istanbul ignore next */ {
    handleError(err, 'app/helpers/renderHTML()')

    return null
  }
}
