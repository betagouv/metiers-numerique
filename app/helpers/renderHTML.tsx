import { handleError } from '@common/helpers/handleError'
import React from 'react'
import rehypeParse from 'rehype-parse'
import rehypeReact from 'rehype-react'
import styled from 'styled-components'
import { unified } from 'unified'

import type { JSXElementConstructor, ReactElement } from 'react'

const HTMLContent = styled.div`
  p,
  li {
    font-size: 1rem;
    line-height: 1.5rem;
  }
`

export function renderHTML(HTMLSource: string): ReactElement<unknown, string | JSXElementConstructor<any>> | null {
  try {
    const htmlContent = unified()
      .use(rehypeParse)
      .use(rehypeReact, { createElement: React.createElement })
      .processSync(HTMLSource).result

    return <HTMLContent className="HTMLContent">{htmlContent}</HTMLContent>
  } catch (err) /* istanbul ignore next */ {
    handleError(err, 'app/helpers/renderHTML()')

    return null
  }
}
