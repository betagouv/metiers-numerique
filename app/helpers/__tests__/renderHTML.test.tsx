/**
 * @jest-environment jsdom
 */
import { renderHTML } from '../renderHTML'

describe.only('app/helpers/renderMarkdownOrHtml()', () => {
  test(`with an HTML source`, () => {
    const markdownSource = [
      '<p>A paragraph.</p>',
      '<p>A <strong>bold</strong> and <em>italic</em> text with a <a href="https://example.org">link</a>.</p>',
      '<ul>',
      '  <li>A</li>',
      '  <li>bullet,</li>',
      '  <li>list.</li>',
      '</ul>',
      '<ol>',
      '  <li>An</li>',
      '  <li>ordered,</li>',
      '  <li>list.</li>',
      '</ol>',
    ].join('\n')

    const result = renderHTML(markdownSource)

    expect(result).toMatchSnapshot()
  })
})
