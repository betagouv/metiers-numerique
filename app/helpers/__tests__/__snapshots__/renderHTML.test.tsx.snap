// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/helpers/renderMarkdownOrHtml() with an HTML source 1`] = `
<styled.div
  className="HTMLContent"
>
  <div>
    <html>
      <head />
      <body>
        <p>
          A paragraph.
        </p>
        

        <p>
          A 
          <strong>
            bold
          </strong>
           and 
          <em>
            italic
          </em>
           text with a 
          <a
            href="https://example.org"
          >
            link
          </a>
          .
        </p>
        

        <ul>
          
  
          <li>
            A
          </li>
          
  
          <li>
            bullet,
          </li>
          
  
          <li>
            list.
          </li>
          

        </ul>
        

        <ol>
          
  
          <li>
            An
          </li>
          
  
          <li>
            ordered,
          </li>
          
  
          <li>
            list.
          </li>
          

        </ol>
      </body>
    </html>
  </div>
</styled.div>
`;
